src/rules/style/end_statements.rs

```diff
@@ -0,0 +1,220 @@
+use crate::ast::FortitudeNode;
+use crate::settings::Settings;
+use crate::{ASTRule, Rule, Violation};
+use tree_sitter::Node;
+
+pub struct AbbreviatedEndStatement {}
+
+impl Rule for AbbreviatedEndStatement {
+    fn new(_settings: &Settings) -> Self {
+        Self {}
+    }
+
+    fn explain(&self) -> &'static str {
+        "
+        End statements should specify what kind of construct they're ending, and the
+        name of that construct. For example, prefer this:
+
+        ```
+        module mymodule
+          ...
+        end module mymodule
+        ```
+
+        To this:
+
+        ```
+        module mymodule
+          ...
+        end
+        ```
+
+        Or this:
+
+        ```
+        module mymodule
+          ...
+        end module
+        ```
+
+        Similar rules apply for submodules, programs, functions, subroutines, and
+        derived types.
+        "
+    }
+}
+
+fn map_start_statement(kind: &str) -> &'static str {
+    match kind {
+        "module" => "module_statement",
+        "submodule" => "submodule_statement",
+        "program" => "program_statement",
+        "function" => "function_statement",
+        "subroutine" => "subroutine_statement",
+        "module_procedure" => "module_procedure_statement",
+        "derived_type_definition" => "derived_type_statement",
+        _ => unreachable!("Invalid entrypoint for AbbreviatedEndStatement"),
+    }
+}
+
+fn map_end_statement(kind: &str) -> &'static str {
+    match kind {
+        "module" => "end_module_statement",
+        "submodule" => "end_submodule_statement",
+        "program" => "end_program_statement",
+        "function" => "end_function_statement",
+        "subroutine" => "end_subroutine_statement",
+        "module_procedure" => "end_module_procedure_statement",
+        "derived_type_definition" => "end_type_statement",
+        _ => unreachable!("Invalid entrypoint for AbbreviatedEndStatement"),
+    }
+}
+
+fn map_statement_type(kind: &str) -> &'static str {
+    match kind {
+        "module" => "module",
+        "submodule" => "submodule",
+        "program" => "program",
+        "function" => "function",
+        "subroutine" => "subroutine",
+        "module_procedure" => "procedure",
+        "derived_type_definition" => "type",
+        _ => unreachable!("Invalid entrypoint for AbbreviatedEndStatement"),
+    }
+}
+
+impl ASTRule for AbbreviatedEndStatement {
+    fn check<'a>(&self, node: &'a Node, src: &'a str) -> Option<Vec<Violation>> {
+        let kind = node.kind();
+        let end_kind = map_end_statement(kind);
+        let end_node = node.child_with_name(end_kind)?;
+
+        // If end node is named, move on.
+        // Not catching incorrect end statement name here, as the compiler should
+        // do that for us.
```

Comment:
I do think that would be a good (separate) warning though -- if we can do easy compiler warnings, we can catch them immediately as people write them (and apply autofixes!)

Reply:
Good point, when we get around to language server bindings that sort of error would be good to catch.